Implement Dispose for types which have disposable data

FiniteReality opened the pull request:
This is likely a **large breaking change** for some people, so should be merged before 2.0 is released.

I still need to handle Dispose warnings for these:
- [x] Discord.Net.Websocket/Entities/SocketGuild
- [x] Discord.Net.Tests

None of this has been thoroughly tested, either. I'm just opening the PR to gather feedback on things I might be missing as I work on the PR.

This should fix #1168 and #1084.

FiniteReality committed:
Initial set of dispose implementations

Not handled yet:
- Discord.Net.Websocket/Entities/SocketGuild
- Discord.Net.Tests

FiniteReality committed:
Refactor DiscordSocketClient init into ctor

This way we remove an IDisposableAnalyzer warning for not disposing
the client when we set the client variable.

AntiTcb commented on src/Discord.Net.WebSocket/DiscordShardedClient.cs:
```diff
@@ -369,5 +375,20 @@ Task<IReadOnlyCollection<IVoiceRegion>> IDiscordClient.GetVoiceRegionsAsync(Requ
         Task<IVoiceRegion> IDiscordClient.GetVoiceRegionAsync(string id, RequestOptions options)
             => Task.FromResult<IVoiceRegion>(GetVoiceRegion(id));
+
+        internal override void Dispose(bool disposing)
```
Should this go through each shard and dispose those as well?

FiniteReality replied:
Yeah, it should. I missed another thing :sweat_smile:

FiniteReality committed:
Dispose of clients when disposing sharded client

FiniteReality commented on src/Discord.Net.WebSocket/DiscordShardedClient.cs:
```diff
@@ -369,5 +375,22 @@ Task<IReadOnlyCollection<IVoiceRegion>> IDiscordClient.GetVoiceRegionsAsync(Requ
         Task<IVoiceRegion> IDiscordClient.GetVoiceRegionAsync(string id, RequestOptions options)
             => Task.FromResult<IVoiceRegion>(GetVoiceRegion(id));
+
+        internal override void Dispose(bool disposing)
```
Would it be a good idea to call StopAsync in Dispose here? We do it in other places (e.g. DiscordSocketClient)

AntiTcb replied:
Sharded client's stop async just calls stop async for each shard, so if their disposers call StopAsync, it's enough just to dispose them.

brianebeling commented on samples/01_basic_ping_bot/Program.cs:
```diff
@@ -14,23 +14,31 @@ namespace _01_basic_ping_bot
     // - https://github.com/foxbot/DiscordBotBase - a bare-bones bot template
     // - https://github.com/foxbot/patek - a more feature-filled bot, utilizing more aspects of the library
-    class Program
+    class Program : IDisposable
```
Is that really necessary? On termination of the program that is already taken care of.

FiniteReality replied:
This is simply not true. For example, if you use a Timer and do not dispose it, your application may not close properly if it is ran using `dotnet run`.

brianebeling commented on samples/01_basic_ping_bot/Program.cs:
```diff
@@ -14,23 +14,31 @@ namespace _01_basic_ping_bot
         static void Main(string[] args)
-            => new Program().MainAsync().GetAwaiter().GetResult();
+        {
+            using (var program = new Program())
```
Then we also would not need this anymore and could go back to an async Main method or get rid of the using at least.

FiniteReality replied:
The socket client should be disposed to prevent memory leaks.

RethoughtSolutions commented on samples/01_basic_ping_bot/Program.cs:
```diff
@@ -14,23 +14,31 @@ namespace _01_basic_ping_bot
+        }
 
-        public async Task MainAsync()
+        public Program()
```
In addition I would advice against doing this kind of work in the constructor. Despite it causing confusion or rather being "untypically". Complex work in constructors should be avoided because of awkward behaviour when throwing exceptions. That is exactly why creational patterns exist. 
In this case there (probably) will never be an exception thrown. But for readability reasons I would just move that into MainAsync and be done with it.

brianebeling commented on samples/01_basic_ping_bot/Program.cs:
```diff
@@ -14,23 +14,31 @@ namespace _01_basic_ping_bot
+        }
 
-        public async Task MainAsync()
+        public Program()
```
In addition I would advice against doing this kind of work in the constructor. Despite it causing confusion or rather being "untypically". Complex work in constructors should be avoided because of awkward behaviour when throwing exceptions. That is exactly why creational patterns exist. 
In this case there (probably) will never be an exception thrown. But for readability reasons I would just move that into MainAsync and be done with it.

FiniteReality replied:
As far as I'm aware, lots of initializer work has been done in constructors, so this is fine. And if the constructor throws (which I doubt it will), we're likely in a state where we cannot logically continue anyway, so for a sample, this is fine.

FiniteReality committed:
Finish implementing IDisposable where appropriate

I opted to use NoWarn in the Tests project as it wasn't really necessary
considering that our tests only run once

FiniteReality commented:
Okay, so I've finished the first pass I believe. This should grab everything that had disposable data properly, but I would like some help testing it

FiniteReality committed:
Tweak samples after feedback

FiniteReality commented:
This should be more or less ready to review now. cc @foxbot
